### ISSUE TITLE
Batch verification, extraction and deletion of files problem

### ISSUE THREAD

**Issue by MicroB9** [Yutaka-Sawada/MultiPar#59]
I am trying to use MultiPar for batch file multiple par archives repair & extraction and deletion of files after successful repair, but I am running into issues with lack of documentation I was hoping to get help with. So far I tried:

1.	running MultiPar from batch file. The first incomplete par set it finds will halt MultiPar from progression to the next par file set. How do I make MultiPar exit even on repair error? Without it, it makes running a batch file impossible.

2.	ParQueue works except it does not delete extracted files after successful recovery. MultiPar is set to do this and it works, but how do I get ParQueue to also follow this MultiPar setting as well instead of ignoring it?

3.	Par2j64.exe – what is the switch option to enable deleting files after successful recovery? I have not been able to figure out the command switch to make it happen here.

Thanks for any help.

**Comment by Yutaka-Sawada** [comment]
> 1. running MultiPar from batch file.
> How do I make MultiPar exit even on repair error?

I don't know how is the batch file. I suspect that the batch script waits the finish of MultiPar.exe. But, MultiPar.exe doesn't close automatically, when it failed to repair. This behavior was made to confirm the result of verification or repair. At this time, you need to close MultiPar.exe manually. If you want to close automatically at failed repair, I may add another setting in MultiPar.ini file. But, you need to edit the .INI file by a text edior. (You cannot change setting on GUI Option window.)

> how do I get ParQueue to also follow this MultiPar setting as well instead of ignoring it?

I didn't implement the task, because nobody requested. As I am lazy, I will add a feature, only when users want. Do you want to enable "Delete recovery files after Repair" option in ParQueue ? If I implement the feature in ParQueue.exe, it will delete all PAR2 files belong to the base name. It will determine by the filename only, not by their status. There is a risk of deleting different PAR2 files, if their filenames are similar.

> Par2j64.exe – what is the switch option to enable deleting files after successful recovery?

I think that you say about deleting PAR2 files. par2j doesn't have such feature, because it's difficult to determine which PAR2 files should be deleted. MultiPar.exe recognizes each PAR2 file's state and deletes.

**Comment by Yutaka-Sawada** [comment]
> I am trying to use MultiPar for batch file multiple par archives repair & extraction and deletion of files after successful repair,

By checking an option: "Delete recovery files after Repair", MultiPar will delete PAR2 files after successful repair. But, it doesn't delete PAR2 files, if it verified all source files were complete already at first. This behavior is because a user may create PAR2 files with MultiPar. If it deletes PAR2 files after verification, a user cannot test the PAR2 files after creation.

Now, I have a question about ParQueue's possible behavior. Should ParQueue delete PAR2 files, when it verifies all source files are complete ? Or, should it keep PAR2 files at verification as same as MultiPar does ? I'm not sure how is your usage.

**Comment by MicroB9** [comment]
Hi,

 

Thanks for reply. Here is what I am looking for. I have a folder with a bunch of subfolders. Each subfolder contains many different par files, and these par files sometimes do not exactly match their extraction filenames.

 

Right now, I wrote a batch file using par2j64 that will check every .par file in all subfolders and if the returned code is successful repair or no need to repair, it will then delete the given .par file.

 

This approach is not working 100% and is rather slow.

 

 

I think MultiPar could work if there was an option I could set to

 

Close MultiPar even after errors, i.e. unable to repair, so it doesn’t hang my batch file, and

Delete damaged and recovery files after successful repair OR if no repair was needed, and

Optionally, keep every new instance of MultiPar in the background so it does not steal focus, otherwise the PC would be unusable while the batch file is opening & closing multipar many times in a row.

 

 

Would this be possible?

 

 

What is important here however is that par files to be deleted are determined from the .par record and not just by matching their filename only, as this could result in data loss for me, since some par files do not match their filenames.

 

 

You may also want to know that did try ParQueue  a few times but each time it hanged permanently maybe after doing 5-10 pars. I find Multipar to be much more stable.

 

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
Hi,

 

I think there should be an option in MultiPar to delete PAR2 (and PAR) files not only after it verifies all source files are complete, regardless if they needed repair or not. I understand this option cannot be turned on when creating par files but if you are doing a batch extract, this option could be very useful.

 

Since ParQueue hangs all the time for me I do not find it usable for me at this time.

 

 

 

Rudolf Galan

**Comment by Yutaka-Sawada** [comment]
> Would this be possible?

I added a new command for MultiPar.exe. It's "/batch", which may replace "/verify". It will open Verify window in minimized state. It will delete recovery files at verification, when all source files are complete. It will close Verify window, even when it failed to repair.

I put the sample (gui_sample_2022-01-29.zip) in "MultiPar_sample" folder on [OneDrive](https://1drv.ms/u/s!AtGhNMUyvbWOaSo1n_R8awJ_hg0?e=4V0gXu). Please test the behavior.

> You may also want to know that did try ParQueue  a few times but each time it hanged permanently maybe after doing 5-10 pars.

I don't know why it failed. Because it doesn't show progress at all, it's difficult to see what happened. Maybe logging output would help. I will solve the problem, when there are more information.

**Comment by MicroB9** [comment]
Hi,

 

I ran some quick tests. Everything seems to be working as expected. The only glitch I found is when I invoked MultiPar on a .par file instead of .par2. A file I tested with had a recovery block missing and MultiPAR in batch mode failed to auto close, after correctly showing blocks missing on this repair.

 

Can you extend the same /batch behavior to .par files as well? That means deleting all .par files after successful repair even if no repair is needed and autoclosing multipar regardless of success or error result?

 

 

As far as ParQueue, I prefer my batch file with multipar.exe /batch. If you have a way of sharing batch files with other users, I would be happy to submit my batch file to you so others can use it with MultiPar also.

 

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
I am running more tests and I just came across a .par2 file, where MultiPAR says need 59 more blocks, but it fails to close its own window. This means my batch script just hangs on MultiPAR window that fails to close. Any ideas? This already happens on all .par files but this is the first time I see this behavior with a .par2 file as well.

 

 

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
I think I was able to resolve the hangs on .par2 file by disabling in the options “monitor files to verify”. As of now, MultiPAR once again only hangs on .par files.

 

It would be nice if /batch switch also forced certain options such as no “monitor files to verify”, yes to “repair automatically”, yes to “after repair, close automatically”, etc. Without manually setting these options, /batch switch may not perform for users as they would expect it to work.

 

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
Another thing, when multipar is opening and closing par2 files, from my

a batch file, despite running minimized, it still manages to steal focus from some applications. This makes the PC not usable while MultiPar batch is running. Is it perhaps possible to run it hidden altogether to avoid this focus stealing?

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
MultiPAR just failed to close itself again. It says “The specified file is not valid” and then it does not close itself after that.

 

 

 

 

Rudolf Galan

**Comment by MicroB9** [comment]
I am seeing another problem. My Parj64 batch that I wrote would check every .par2 file even if it did not need repair. That’s why I assumed that MultiPar batch file would be several times in magnitude faster. But I am actually seeing the opposite. My ParJ64 batch is about 3 times faster than MultiPAR batch, despite MultiPAR deleting no longer needed par2 files.

 

I am assuming this differece in speed is due to ParJ64 being compiled differently than MultiPAR? Any chance you could add parj64 command line switch to “delete recovery files after successful recovery or if no recovery is needed”? I think this would be the fastest solution for me.

 

 

 

 

Rudolf Galan

**Comment by Yutaka-Sawada** [comment]
> A file I tested with had a recovery block missing and MultiPAR in batch mode failed to auto close, after correctly showing blocks missing on this repair.
> Can you extend the same /batch behavior to .par files as well?

 There is no difference between treating PAR1 and PAR2. PAR1 works on my PC, too. I don't know why it did not work on your PC. You had better check your command-line.

> If you have a way of sharing batch files with other users, I would be happy to submit my batch file to you so others can use it with MultiPar also.

There is a page of batch scripting on Help documents. The title is "Batch script" and filename is "batch.htm". You may see the page with Internet Browser. If your batch file isn't so large, I may put the script as a sample on the help.

If you want to share information with other GitHub users, you may use [Discussions page](https://github.com/Yutaka-Sawada/MultiPar/discussions). The page is more stable than this issue page.

> It would be nice if /batch switch also forced certain options such as no “monitor files to verify”, yes to “repair automatically”, yes to “after repair, close automatically”, etc.

 This is a nice idea. I implemented the way to change options temporary. A user can select which option to change by adding a line on MultiPar.ini manually. By default, 3 options are changed temporary with "/batch" command;
It forces disable option "Monitor files to verify".
It forces enable option "Repair automatically" after Verification.
It forces enable option "Close automatically" after Verification or Repair.

> despite running minimized, it still manages to steal focus from some applications.
> Is it perhaps possible to run it hidden altogether to avoid this focus stealing?

Hide window is bad, because a user cannot treat error. I changed the window state from SW_MINIMIZE to SW_SHOWMINNOACTIVE. This may solve problem.

> It says “The specified file is not valid” and then it does not close itself after that.

The PAR file on the script might be deleted already. To check error, MultiPar won't close when there was a error.

> My ParJ64 batch is about 3 times faster than MultiPAR batch
> I am assuming this differece in speed is due to ParJ64 being compiled differently than MultiPAR?

MultiPar 's automated behavior may wait a few seconds between each task. This waiting is same on ParQueue, too. The slowness is good to decrease stress on hardware.

> Any chance you could add parj64 command line switch to “delete recovery files after successful recovery or if no recovery is needed”?

I added an option "/p" to purge recovery files. It will send recovery files into recycle bin, when all source files are complete after verification or repair. I changed a way of using recycle bin from SHFileOperation to IFileOperation for Windows Vista or later. Now, it seems to support long path.

I put the sample (MultiPar_sample_2022-01-30.zip) in "MultiPar_sample" folder on [OneDrive](https://1drv.ms/u/s!AtGhNMUyvbWOaSo1n_R8awJ_hg0?e=4V0gXu). Please test the behavior. Because I had no time to test so much, there may be a bug or failure still.

**Comment by MicroB9** [comment]
I have tested parj64 with the /p switch and it appears to work as I noticed some of par files disappeared from the queue, i.e. they were delete by parj64. I think that is the best solution I can hope for.

 

 

 

 

Rudolf Galan
